### DISCUSSION TITLE
Client error Interceptors

### DISCUSSION THREAD

**Discussion by danijel-tolj** [appwrite/appwrite#2482] [Ideas]
I have been using Appwrite with Flutter for a couple of days, and I noticed a glaring issue that should be rather easy to implement. 
It is the lack of flexibility when handling errors. For example, if you receive a 401, it would be nice to have an error interceptor that handles that case globally and signs you out of the app. 
Right now, the solution for that would be calling account.get() to try and retrieve a session, but I find that too cumbersome and a suboptimal replacement for a simple error interceptor. What are your thoughts on this ? 

My proposal would be to have something like this : 

class MyInterceptor extends ErrorInterceptor{
  @override
  FutureOr<dynamic> onError(AppwriteException exception) {
    if (exception.code==401) {
    // Sign user out of the app 
    }
    return exception;
   }
  }

**Comment by lohanidamodar** [comment]
This looks like a good enhancement we can add to the SDKs. I like the idea.

**Reply by eldadfux** [reply]
How will it work with Appwrite built-in exceptions?

**Reply by lohanidamodar** [reply]
Same way it works now, just that our client will accept error interceptors and we check and apply the interceptors when exceptions are received before throwing the exception.

**Comment by maeddin** [comment]
The feature would be nice, but not urgent. For example, I have written a simple wrapper that takes care of this. Of course you always have to pass the function call to this wrapper, but this is not very complex and relatively flexible, if you want to quickly add edge cases, for example. In addition, you make each call to AppWrite in the best case anyway only at one point in the code.
But of course it would still be an improvement that I support.

**Reply by lohanidamodar** [reply]
Great point @MaddinMade, we can always use wrapper. And like you I believe this is a nice to have addition to our SDKs.

**Comment by abnegate** [comment]
I like the idea. I think we could expand on it to include generalised interceptors, where they could be applied to the request, response or error.  Then developers could add an error interceptor as above, but could also add other custom interceptors like logging, analytics, retry handlers etc. It would be helpful for debugging issues too.

**Comment by eldadfux** [comment]
@abnegate @lohanidamodar do you have any code examples of how you would imagine this functionally in the client object?

**Comment by abnegate** [comment]
@eldadfux Something like so in my mind:

## SDK Code
```kotlin
internal interface Interceptor<T> {
    fun intercept(obj: T): T
}
interface RequestInterceptor: Interceptor<Request> {
    fun intercept(request: Request) : Request
}
interface ErrorInterceptor: Interceptor<Throwable> {
    fun intercept(throwable: Throwable) : Throwable
}
```

### Client Additions

```kotlin
val interceptors: MutableList<Interceptor<*>>

fun addInterceptor(interceptor: Interceptor) {
    interceptors.add(interceptor)
}
```

```kotlin
// ...
// On error thrown
catch(ex: Exception) {
    interceptors
        .find { it is ErrorInterceptor }
        .forEach { it.intercept(ex) }
}
// ...
```

## Developer Code

```kotlin
class MyErrorInterceptor : ErrorInterceptor {
    fun intercept(throwable: Throwable): Throwable {
        print("Encountered error: ${throwable.message}")
        return throwable
    }
}
```

```kotlin
client.addInterceptor(MyErrorInterceptor());
```
